framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -339,106 +339,126 @@ public TransferFunction createFlowTransferFunction(CFAbstractAnalysis<Value, Sto
+     * the standard defaults.
+     *
+     * @param defs QualifierDefault object to which defaults are added
+     */
+    protected void addUncheckedCodeDefaults(QualifierDefaults defs) {
+        for (Class<? extends Annotation> annotation : getSupportedTypeQualifiers()) {
+            DefaultInUncheckedCodeFor defaultForUnannotated = annotation.getAnnotation(DefaultInUncheckedCodeFor.class);
+
+            if (defaultForUnannotated != null) {
+                final DefaultLocation[] locations = defaultForUnannotated.value();
+                defs.addUncheckedCodeDefaults(AnnotationUtils
+                                                      .fromClass(elements,
+                                                                        annotation),
+                                                     locations);
             }
-            Set<? extends AnnotationMirror> bottoms = this.qualHierarchy.getBottomAnnotations();
-            for (AnnotationMirror bot : bottoms) {
-                defs.addUncheckedCodeDefault(bot, DefaultLocation.PARAMETERS);
-                defs.addUncheckedCodeDefault(bot, DefaultLocation.LOWER_BOUNDS);
-                defs.addUncheckedCodeDefault(bot, DefaultLocation.FIELD);
-                // TODO: this isn't simply DefaultLocation.OTHERWISE, because that would
-                // also apply to type declarations, which isn't currently working as desired.
-                // See: https://groups.google.com/d/msg/checker-framework-dev/vk2V6ZFKPLk/v3hENw-e7gsJ
+
+            if (annotation.getAnnotation(DefaultQualifierInHierarchyInUncheckedCode.class) != null) {
+                if (defaultForUnannotated != null) {
+                    // A type qualifier should either have a DefaultForUncheckedCode or
+                    // a DefaultQualifierForUncheckedCode annotation.
+                    ErrorReporter
+                            .errorAbort("GenericAnnotatedTypeFactory.createQualifierDefaults: " +
+                                                "qualifier has both @DefaultForUncheckedCode and @DefaultQualifierInHierarchyForUncheckedCode annotations: "
+                                                + annotation.getCanonicalName());
+                } else {
+                    defs.addUncheckedCodeDefault(AnnotationUtils.fromClass(elements, annotation),
+                                                      DefaultLocation.OTHERWISE);
+                }
             }
         }
+        Set<? extends AnnotationMirror> tops = this.qualHierarchy.getTopAnnotations();
+        Set<? extends  AnnotationMirror> bottoms = this.qualHierarchy.getBottomAnnotations();
+        defs.addUncheckedStandardDefaults(tops, bottoms);
 
-        return defs;
+        // Don't require @DefaultQualifierForUncheckedCode or an unchecked default for DefaultLocation.OTHERWISE.
+        // if one isn't specified the defaults for checked code will be used.
```

Comment:
Could you please add this note to the javadoc of both DefaultInUncheckedCodeFor and DefaultQualifierInHierarchyInUncheckedCode?